src/cli/em_cli.cpp

```diff
@@ -395,21 +415,132 @@ em_network_node_t *em_cli_t::get_network_tree(char *buff)
 {
 	cJSON *obj = NULL;
 	em_network_node_t *root;
+	unsigned int node_display_ctr = 0;
 
 	if ((obj = cJSON_Parse(buff)) == NULL) {
 		return NULL;
 	}
 
 	root = (em_network_node_t *)malloc(sizeof(em_network_node_t));
 	memset(root, 0, sizeof(em_network_node_t));
-	get_network_tree_node(obj, root);
+	
+	get_network_tree_node(obj, root, &node_display_ctr);
 
 	cJSON_Delete(obj);
 
 	return root;
 
 }
 
+em_network_node_t *em_cli_t::get_node_from_node_ctr(em_network_node_t *tree, unsigned int node_display_ctr)
+{
+	em_network_node_t *node = NULL;
+	bool found_match = false;
+	unsigned int i;
+	em_long_string_t	dbg;
+
+	snprintf(dbg, sizeof(em_long_string_t), "Node: %s, Node Counter: %d", tree->key, tree->display_info.node_ctr); 
+	dump_lib_dbg(dbg);	
+
+	if (tree->display_info.node_ctr == node_display_ctr) {
+		return tree;
+	} else {
+		for (i = 0; i < tree->num_children; i++) {
+			if ((node = get_node_from_node_ctr(tree->child[i], node_display_ctr)) != NULL) {
+				found_match = true;
+				break;
+			}
+		}
+
+		if (found_match == true) {
+			return node;
+		}
+	}
+
+	return NULL;
+}
+
+em_network_node_t *em_cli_t::clone_network_tree(em_network_node_t *orig_node, em_network_node_t *dis_node, unsigned int index, bool collapse, unsigned int *node_display_ctr)
+{
+	em_network_node_t *cloned = NULL, *tree_to_add = NULL;
+	unsigned int i;
+	bool should_consider = false;
+	em_network_node_t *node;
+	unsigned int start_node_ctr = 0;
+
+	if (node_display_ctr == NULL) {
+		node_display_ctr = &start_node_ctr;
+	}
+
+	if (orig_node == NULL) {
+		return NULL;
+	}
+
+	if (dis_node == NULL) {
+		node = orig_node;
+	} else {
+		node = dis_node;
+	}
+
+	cloned = (em_network_node_t *)malloc(sizeof(em_network_node_t));
+	memset(cloned, 0, sizeof(em_network_node_t));
+	
+	strncpy(cloned->key, node->key, strlen(node->key) + 1);
+	memcpy(&cloned->display_info, &node->display_info, sizeof(em_node_display_info_t));
+	cloned->display_info.node_ctr = *node_display_ctr;
+	cloned->display_info.orig_node_ctr = node->display_info.orig_node_ctr;
+
+	cloned->type = node->type;
+	strncpy(cloned->value_str, node->value_str, strlen(node->value_str) + 1);
+	cloned->value_int = node->value_int;
+
+	should_consider = (node->display_info.node_ctr == index);
+		
+	if (should_consider == true) {
+		if (collapse == false) {
+			tree_to_add = get_node_from_node_ctr(orig_node, node->display_info.orig_node_ctr);	
+			dump_lib_dbg(get_network_tree_string(tree_to_add));
```

Comment:
Since get_network_tree_string is returing a malloced string, I think the return value should be freed after dump_lib_dbg call